### PR TITLE
Errors handler in CE

### PR DESCRIPTION
**Pull request by SouroucheB**
capture du 2018-02-13 11-56-23

### CODE REVIEW

File: lib/router.js [hunk 1]
```diff
@@ -282,7 +282,7 @@ module.exports = class Router extends express.Router {
           res.end('<script>window.close();</script>');
         })
         .catch((err) => {
-          res.status(SYSTEM_ERROR_STATUS).send(err);
+          res.end(`<script>window.opener.postMessage('${JSON.stringify(err)}', '*'); window.close();</script>`);
```
Comment (SouroucheB) [inline lib/router.js]: Mozilla tells that providing "*" for domain parameter is a very bad practice but I did not know what I could provide in place. A complete URI like Paht.join(dirname, ...) ?

File: src/js/CloudExplorer.jsx [hunk 2]
```diff
@@ -47,6 +47,9 @@ export default class CloudExplorer extends React.Component {
    */
 
   componentDidMount () {
+    window.addEventListener('message', (err) => {
```
Comment (lexoyo) [inline src/js/CloudExplorer.jsx]: this should probably be in UnifileService, where the window is created

File: src/js/CloudExplorer.jsx [hunk 3]
```diff
@@ -47,6 +47,9 @@ export default class CloudExplorer extends React.Component {
...
+      this.onUnifileError(JSON.parse(err.data));
```
Comment (lexoyo) [inline src/js/CloudExplorer.jsx]: no remove event?

File: src/js/CloudExplorer.jsx [hunk 4]
```diff
@@ -47,6 +47,9 @@ export default class CloudExplorer extends React.Component {
    */
 
   componentDidMount () {
```
Comment (lexoyo) [inline src/js/CloudExplorer.jsx]: every time the component is created it will add a listener

File: lib/router.js (hunk 1 above)
Comment (lexoyo) [reply lib/router.js]: you can provide the domain we are on

File: src/js/UnifileService.js [hunk 5]
```diff
@@ -154,11 +154,17 @@ export default class UnifileService {
...
+    reject(new Error(JSON.parse(error.data).message));
+  }
+
   // The auth method has to be called on a click or keydown in order not to be blocked by the browser
   auth (service) {
     return new Promise((resolve, reject) => {
       // Open a blank window right away, before we know the URL, otherwise the browser blocks it
       const win = window.open();
+      window.addEventListener('message', (error) => this.rejectError(reject, error));
+      window.removeEventListener('message', (error) => this.rejectError(reject, error));
```
Comment (lexoyo) [inline src/js/UnifileService.js]: this line does nothing

File: src/js/UnifileService.js [hunk 6]
```diff
@@ -154,11 +154,17 @@ export default class UnifileService {
...
+  rejectError (reject, error) {
+    reject(new Error(JSON.parse(error.data).message));
+  }
+
   // The auth method has to be called on a click or keydown in order not to be blocked by the browser
   auth (service) {
     return new Promise((resolve, reject) => {
       // Open a blank window right away, before we know the URL, otherwise the browser blocks it
       const win = window.open();
+      window.addEventListener('message', (error) => this.rejectError(reject, error));
```
Comment (lexoyo) [inline src/js/UnifileService.js]: if this happens, "startPollingAuthWin" will still call "resolve"

Comment (SouroucheB) [reply src/js/UnifileService.js]: Holly shit, I've just forgot the return word !

File: src/js/UnifileService.js [hunk 7]
```diff
@@ -159,6 +159,11 @@ export default class UnifileService {
     return new Promise((resolve, reject) => {
       // Open a blank window right away, before we know the URL, otherwise the browser blocks it
       const win = window.open();
+      const rejectError = (error) => {
+        window.removeEventListener('message', rejectError);
+        return reject(new Error(JSON.parse(error.data).message));
```
Comment (lexoyo) [inline src/js/UnifileService.js]: do not return anything here

Comment (lexoyo) [reply src/js/UnifileService.js]: `reject` and `resolve` will still be called by `authEnded`

**Comment by lexoyo** [comment]
Feel free to reopen this if you want to contribute
Closing for now

**Pull request closed without merging**